### ISSUE TITLE
remove halo exchanges for gh_inc dependencies

### ISSUE THREAD

**Issue by rupertford** [stfc/PSyclone#158]
Halo exchanges are currently added unnecessarily for gh_inc accesses.

If we ignore the effect of redundant computation transformations then, when a field is computed with gh_inc, it will be computed into the level 1 halo. This is performed to ensure that the fields owned and annexed dofs at the edge of the domain are computed correctly. At the moment it is assumed that the fields values must be correct within the halo and therefore a halo exchange will be added before the computation. However this is not the case. Only the owned and annexed dofs need to be valid. This mistake was pointed out by @thomasmelvin

**Comment by TeranIvy** [comment]
Added myself to keep track of this interesting issue!

**Comment by rupertford** [comment]
Created gh_inc_nohex branch

**Comment by rupertford** [comment]
I'm going to summarise the new logic needed here

A gh_inc access that loops to the l1 halo only requires owned and annexed dofs to be up to date. If the loops upper bound is ln then l(n-1) must be up to date.

If we compute annexed dofs then the annexed dofs for gh_inc will always be correct. Therefore we do not need a halo exchange for a gh_inc access with loop bounds to l1 halo.

If we do not compute annexed dofs then we will need a halo exchange if we know that annexed dofs are invalid or if we are not sure, otherwise we do not need a halo exchange.

In both cases an access up to level n>1 halo will require correct halo data to level n-1 and will result in clean halos to level n-1.

**Comment by rupertford** [comment]
The halo exchange logic for writers has the idea of the outermost halo being clean/dirty after redundant computation (dirty_outer). I've added halo exchange logic for readers that specifies whether the outermost halo needs to be clean or not (clean_outer). It does not when the read access is gh_inc.

**Comment by rupertford** [comment]
I needed to modify a load of tests as the number of halo exchanges produced has now changed and this affects a number of existing tests when COMPUTE_ANNEXED_DOFS is True as there is now no halo exchange required before a gh_inc access when the previous dependence is not known. All tests now pass with COMPUTE_ANNEXED_DOFS being defaulted to True or False (currently the default is False) and with or without compilation. Most of the modified tests now check for COMPUTE_ANNEXED_DOFS = False or True, irrespective of the default, using a fixture.

**Comment by rupertford** [comment]
Hi @TeranIvy. This branch is now up-to-date with master. You kindly offered to test whether we get the same results when we use it with LFRic. If it is working correctly there should be no need to make any changes to the existing code and the results should bit compare. If it does bit compare then I'm afraid it will still need testing a second time with annexed_dofs set to True (it is still False by default). In this latter case the same changes that you made previously would need to be applied to the PSyKAl-lite code. Thanks!

**Comment by TeranIvy** [comment]
Hi @rupertford! I'm in the process of testing the `GH_INC` branch on LFRic trunk. Just `GH_INC` works fine, however there are some issues with annexed dofs set to `True`. Just to confirm - I am running the trunk in parallel on Desktop with just 1 OMP thread as I did the last week. Is this still the correct way to run? I will open a ticket in LFRic with my findings so we can look further into the issues.

**Comment by rupertford** [comment]
Hi @TeranIvy. It's good to hear that the gh_inc change gives correct results.

As for gh_inc + annexed_dofs you are correct in thinking that testing with 1 OMP thread should be valid. The use of OMP threads (or not) should not make any difference to the halo exchange calls. The only thing you will need to do is make the changes to the PSyKAl-lite code as before, as annexed_dofs relies on this (but I've already said that earlier so I'm repeating myself!).

**Comment by TeranIvy** [comment]
Ok, thanks. I have more detailed results of testing the LFRic code with gh_inc and annexed dofs however I do not have time to write them down into the [LFRic ticket #1660](https://code.metoffice.gov.uk/trac/lfric/ticket/1660). The testing methodology is described there, as well as the location of the LFRic branch.

The main failures are:
- 1 test in `gungho` developer Rose suite and 1 `gungho` test in nightlies (both fail the KGO checks);
- `gungho` Desktop application has different outputs for conservation measures in Log files, residuals in Helmholtz solvers and KGO cheksums when run in parallel. The worrying bit is that the global mass (first value in conservation output) is also different and that is calculated from `rho` field which lives on W3 space;
- `gravity_wave` miniapp fails to converge when run in parallel.

I will put more details in the LFRic ticket on Monday.

**Comment by rupertford** [comment]
I guess one possible option might be field values that a read in/initialised outside of PSyclone. If these don't have correct annexed dof values then there could be an issue. However, If that were the case then I would expect the code to also fail with annexed dofs and not gh_inc.

**Comment by TeranIvy** [comment]
I have some good news. It seems that I omitted a few instances of `last_dof_owned` in `psykal_lite_vector_mod.F90` last time I tested this branch. I re-run the tests with the recent trunk and the summary is below.
1) Just `NoHex GH_INC`: everything is the same as before (see [comment 5 in LFRic ticket 1660](https://code.metoffice.gov.uk/trac/lfric/ticket/1660#comment:5)).
2) `NoHex GH_INC + Annexed Dofs`: there are a couple of KGO check failures in Rose suite but all compared Desktop applications give the same results (see [comment 7 in LFRic ticket 1660](https://code.metoffice.gov.uk/trac/lfric/ticket/1660#comment:7)) and everything else seems fine so far. We will do more exploring for different KGOs but I think this PSyclone branch can go to code review.
3) The number of halo exchanges is significantly reduced between the current PSyclone and PSyclones built with `NoHex GH_INC` and `NoHex GH_INC + Annexed Dofs`, respectively (see [comment 8 in LFRic ticket 1660](https://code.metoffice.gov.uk/trac/lfric/ticket/1660#comment:8) for more details).
